Brokerages/Fxcm/FxcmBrokerage.cs

```diff
@@ -0,0 +1,364 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using com.fxcm.external.api.transport;
+using com.fxcm.external.api.transport.listeners;
+using com.fxcm.external.api.util;
+using com.fxcm.fix;
+using com.fxcm.fix.pretrade;
+using com.fxcm.fix.trade;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Orders;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Brokerages.Fxcm
+{
+    /// <summary>
+    /// The FXCM brokerage implementation
+    /// </summary>
+    public partial class FxcmBrokerage : Brokerage, IDataQueueHandler, IGenericMessageListener, IStatusMessageListener
+    {
+        private readonly IOrderProvider _orderProvider;
+        private readonly IHoldingsProvider _holdingsProvider;
+        private readonly string _server;
+        private readonly string _terminal;
+        private readonly string _userName;
+        private readonly string _password;
+
+        private Thread _orderEventThread;
+        private volatile bool _orderEventThreadStopped;
```

Comment:
Where appropriate I prefer to use [CancellationTokenSource](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtokensource(v=vs.110).aspx) instead of explicit usage of `volatile` memory barrier semantics